Feature: UFM component filters

leekelleher opened the pull request:
> [!IMPORTANT]
> **The details in this pull request may be outdated.**
> [Please refer to the official documentation for the latest UFM syntax.](https://docs.umbraco.com/umbraco-cms/reference/umbraco-flavored-markdown)

## Description

To support a "filter" feature to Umbraco-Flavored Markdown (UFM), this pull request adds an `ufmFilter` extension type and a base class (`UmbUfmElementBase`) for the UFM components in order to support reuse the filter logic.

Also in this pull request...

- Syntax for content pickers, `{~value}`, to support document, media and member pickers (both single and arrays), rendering the entity's name as the label
- Adds UFM filter functions (extensions), these are piped within the UFM curly bracket syntax; there is...
  - Fallback, i.e. `{= value | fallback:No items }`
  - Lowercase, i.e. `{= value | lowercase }`
  - Strip HTML, i.e. `{= value | strip-html }`
    - this is compatible with the RTE data structure, e.g. `{ markup, blocks }`
  - Title Case, i.e. `{= value | title-case }`
  - Truncate, i.e. `{= value | truncate:10:false }`
    - this supports multiple arguments: `length: number` and `tail: string` (defaulting to "…", and omitted if `"false"`)
  - Uppercase, i.e. `{= value | uppercase }`
  - Word Limit, i.e. `{= value | word-limit:10 }`
    - this supports a `limit: number` argument

Note, UFM filters can be piped together, e.g. `{= value | strip-html | word-limit:30 }`. The filters will also work for the localization and content picker UFM components, e.g. `{#general_choose | uppercase}`; `{~value | title-case | fallback: No items }`.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## How to test?

For general testing of dynamic labels can be done within Collection Views, Block List/Grid and Document property descriptions.

For specific testing of the filters, e.g. in a Collection View, try adding the piped filter syntax to `{=value}` labels; also try piping filters, e.g. `{= value | lowercase | uppercase | title-case | truncate:4:true}`.

leekelleher committed:
UFM feature additions + refactor

- Re-structures "ufm-components" folder
- Adds `ufmFilter` extension type
- Adds `UmbUfmElementBase` to support reusing filters

leekelleher committed:
Code tidy-up

leekelleher committed:
Adds test for `{~contentPicker}` "ufm-content-name" syntax

leekelleher committed:
Exports `UfmPlugin` and `UfmToken` types

leekelleher committed:
Adds UFM filter extensions/functions

- Fallback
- Lowercase
- Strip HTML
- Title Case
- Truncate
- Uppercase
- Word Limit

leekelleher committed:
Resolved a couple of @sonarcloud issues

leekelleher commented:
> Failed conditions
> ❌ 9.8% Duplication on New Code (required ≤ 3%)

re: @sonarcloud's code duplication issue, this is in a manifest file, of which the duplication of JSON structure is intentional.

iOvergaard commented:
> > Failed conditions
> > ❌ 9.8% Duplication on New Code (required ≤ 3%)
> 
> re: @sonarcloud's code duplication issue, this is in a manifest file, of which the duplication of JSON structure is intentional.

I have added `**/manifests?.ts` to be excluded for further duplication detection

leekelleher committed:
Adds base class for UFM filters

leekelleher committed:
Adds UFM global context

This moves the `UmbExtensionsApiInitializer` code from the `ufm-render`
and `ufm-element-base` components to the context, for optimization.

leekelleher committed:
Merge remote-tracking branch 'origin/main' into v14/feature/ufm-filters

iOvergaard committed:
Merge branch 'main' into v14/feature/ufm-filters

iOvergaard committed:
Merge branch 'main' into v14/feature/ufm-filters

iOvergaard committed:
Merge branch 'main' into v14/feature/ufm-filters

iOvergaard committed:
feat: evaluate the truncate value to make sure its a string and has a length, otherwise the function might return "..." leading other filters to be confused
